### PR TITLE
Quantile Regression: pinball loss for pressure calibration

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replace the standard L1 surface loss on pressure with a **quantile regression (pinball) loss** that predicts q10, q50, and q90 quantiles simultaneously. The q50 (median) prediction is used for MAE evaluation. The tail quantiles (q10, q90) force gradient attention toward the hardest-to-predict regions (suction peaks, separation bubbles) where the model systematically underestimates peak magnitude.

At high-AoA suction peaks, pressure is sharply peaked and the model regresses toward the mean — underestimating the peak. Quantile regression inherently focuses gradient on distribution tails, exactly the nodes the model underpredicts. This is a simple loss change with no architectural complexity.

**Literature**: Quantile regression (Koenker & Bassett, 1978) is a well-validated technique that routinely wins Kaggle competitions with skewed targets. Pinball loss is the asymmetric L1 loss. This is different from heteroscedastic loss (#2284, closed) which predicted variance; quantile regression is distribution-free and more robust.

### Step 1: Add pinball loss function

```python
def pinball_loss(y_pred, y_true, quantile):
    """Asymmetric L1 loss for quantile regression."""
    error = y_true - y_pred
    return torch.max(quantile * error, (quantile - 1) * error).mean()
```

### Step 2: Modify pressure output head

Change the surface pressure output from 1 channel to 3 channels (q10, q50, q90). The final output head Linear layer for pressure goes from `d_hidden → 1` to `d_hidden → 3`.

```python
# In the SRF/AftFoilRefinementHead output layer:
# Before: self.pressure_out = nn.Linear(d_hidden, 1)
# After:  self.pressure_out = nn.Linear(d_hidden, 3)  # q10, q50, q90
```

Keep Ux and Uy outputs unchanged (standard L1 loss, 1 channel each).

### Step 3: Modify loss computation for pressure

```python
if args.quantile_loss:
    p_pred_q10 = surf_pred[:, :, 2]   # pressure q10
    p_pred_q50 = surf_pred[:, :, 3]   # pressure q50 (median, used for eval)
    p_pred_q90 = surf_pred[:, :, 4]   # pressure q90
    p_true = surf_target[:, :, 2]     # true pressure

    loss_p = (0.3 * pinball_loss(p_pred_q10, p_true, 0.1) +
              0.4 * pinball_loss(p_pred_q50, p_true, 0.5) +
              0.3 * pinball_loss(p_pred_q90, p_true, 0.9))
else:
    loss_p = F.l1_loss(surf_pred[:, :, 2], surf_target[:, :, 2])
```

Note: Check exact tensor indices for pressure channel — it may be index 0 if `--pressure_first` is set. Verify which index is pressure in the surface prediction tensor.

### Step 4: Use q50 for evaluation

At eval time, use only the q50 output for MAE calculation:
```python
# During eval, extract q50 pressure prediction
p_eval = surf_pred[:, :, q50_idx]  # the median quantile
```

### Step 5: Add flags

- `--quantile_loss` (bool flag)
- Weights are fixed at 0.3/0.4/0.3 for q10/q50/q90

### Training commands (2 seeds)

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/quantile-s42" --wandb_group quantile-regression-decoder \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --quantile_loss

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/quantile-s73" --wandb_group quantile-regression-decoder \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --quantile_loss
```

## Baseline

**Current best** (PR #2319, 2-seed avg):

| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Quantile Regression: pinball loss for pressure (q10/q50/q90)

**Verdict: Clear negative result.** Quantile regression catastrophically degraded all pressure metrics — 2-3x worse than baseline across all splits.

### Metrics (2-seed avg)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in   | 30.7    | 30.6    | **30.65**  | 11.709   | **+162%** ❌ |
| p_oodc | 17.1    | 17.2    | **17.15**  | 7.544    | **+127%** ❌ |
| p_tan  | 56.3    | 49.0    | **52.65**  | 27.402   | **+92%** ❌ |
| p_re   | 13.2    | 12.5    | **12.85**  | 6.481    | **+98%** ❌ |

### Experiment details

**Commands:**
```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/quantile-s42" --wandb_group quantile-regression-decoder \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --quantile_loss

# Seed 73 — same with --seed 73 and CUDA_VISIBLE_DEVICES=1
```

- **W&B runs:** slz2ynxv (seed 42), 1wg8q2qm (seed 73)
- **Epochs:** ~159 (hit timeout)
- **Peak VRAM:** ~83 GB (s42), ~94 GB (s73) — higher than baseline due to 3-channel pressure output

### Implementation

1. Modified pressure output head from 1 → 3 channels (q10, q50, q90)
2. q50 (median) used for velocity conditioning in pressure-first path
3. q50 used for all evaluation MAE and non-pressure losses
4. Surface pressure loss replaced with weighted pinball loss: 0.3×pinball(q10, 0.1) + 0.4×pinball(q50, 0.5) + 0.3×pinball(q90, 0.9)
5. Hard-node mining uses q50 abs error for thresholding, pinball loss for reweighted loss
6. Quantile spread (q90-q10) logged as `train/quantile_spread_surf`

### What happened — why it failed

The quantile regression approach failed catastrophically for three key reasons:

1. **Loss magnitude mismatch (primary cause):** Each pinball loss term has expected value ~0.5×MAE (regardless of quantile level). With weights 0.3/0.4/0.3, the total surface pressure loss is ~0.5× the original L1 loss. This effectively halves the gradient contribution from surface pressure, causing the model to dramatically underfit on pressure.

2. **Capacity dilution:** The pressure head now outputs 3 channels instead of 1, splitting model capacity across q10/q50/q90. The gradient from q10/q90 quantiles actively competes with q50 accuracy — the model can't simultaneously optimize for the median and for tail-quantile diversity with the same parameters.

3. **Already-optimized baseline:** The current baseline is the result of extensive hyperparameter tuning where the loss balance (vol_loss + surf_weight × surf_loss) is already well-calibrated. Halving the effective surface pressure loss magnitude disrupted this carefully tuned balance.

The hypothesis that pinball loss would "force gradient attention toward hardest-to-predict regions" was incorrect — the asymmetric weighting of the pinball loss at q10/q90 doesn't provide focused pressure-peak gradients; it provides weaker gradients overall because each quantile term is inherently 0.5× the L1 magnitude.

### Suggested follow-ups

- **If revisiting quantile regression:** The loss magnitude issue could be fixed by scaling pinball loss by 2× to match L1 magnitude. However, the capacity dilution problem would remain.
- **Alternative for tail focus:** Direct hard-example mining (already in the baseline) or focal-L1 loss (upweighting high-error nodes without changing the output dimensionality) would be a simpler way to target suction peaks without the multi-output penalty.
- **Uncertainty estimation without quantile loss:** A separate lightweight uncertainty head (predicting log-variance) attached post-hoc wouldn't hurt the main prediction path.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2339 because ADVISOR: Closing — quantile regression catastrophically degraded all pressure metrics (p_in +162%, p_oodc +127%, p_tan +92%, p_re +98%). Root cause well identified: pinball loss has ~0.5× gradient magnitude vs L1, which effectively halved the surface pressure gradient contribution. The 3-channel output also diluted capacity. Good analysis of why it failed — the key insight is that our baseline's existing hard-node mining already provides focused gradients at high-error regions without the multi-output penalty. This direction is exhausted..
